File: src/main/java/org/roylance/yaorm/services/proto/IEntityProtoService.kt
package org.roylance.yaorm.services.proto

import org.roylance.yaorm.models.YaormModel

interface IEntityProtoService {
    val indexDefinition: YaormModel.Index?

    fun createTable(definition: YaormModel.TableDefinition): Boolean
    fun dropTable(definition: YaormModel.TableDefinition): Boolean

    fun createIndex(indexModel: YaormModel.Index, definition: YaormModel.TableDefinition): Boolean
    fun dropIndex(indexModel: YaormModel.Index, definition: YaormModel.TableDefinition): Boolean

    fun createColumn(propertyDefinition: YaormModel.ColumnDefinition, definition: YaormModel.TableDefinition): Boolean
    fun dropColumn(propertyDefinition: YaormModel.ColumnDefinition, definition: YaormModel.TableDefinition): Boolean

    fun bulkInsert(instances: YaormModel.Records, definition: YaormModel.TableDefinition): Boolean
    fun createOrUpdate(entity: YaormModel.Record, definition: YaormModel.TableDefinition): Boolean
    fun create(entity: YaormModel.Record, definition: YaormModel.TableDefinition): Boolean
    fun update(entity: YaormModel.Record, definition: YaormModel.TableDefinition): Boolean

    fun updateWithCriteria(
            newValues: YaormModel.Record,
            whereClauseItem: YaormModel.WhereClause,
            definition: YaormModel.TableDefinition): Boolean
    fun updateCustom(customSql: String): Boolean
    fun delete(id: String, definition: YaormModel.TableDefinition): Boolean
    fun deleteAll(definition: YaormModel.TableDefinition): Boolean

    fun getCount(definition: YaormModel.TableDefinition): Long

    fun getManyStream(n: Int=100000, definition: YaormModel.TableDefinition, streamer: IProtoStreamer)

    fun get(id: String, definition: YaormModel.TableDefinition): YaormModel.Record?

    fun getCustom(customSql: String, definition: YaormModel.TableDefinition): YaormModel.Records
    fun getMany(n: Int=1000, definition: YaormModel.TableDefinition): YaormModel.Records
    fun where(whereClauseItem: YaormModel.WhereClause, definition: YaormModel.TableDefinition): YaormModel.Records
}

File: csharp/Yaorm/Yaorm/Services/IEntityProtoService.cs
using System;
using Org.Roylance.Yaorm.Models;

namespace Yaorm
{
	public interface IEntityProtoService
	{
		Index Index { get; }

		bool CreateTable(TableDefinition tableDefinition);
		bool DropTable(TableDefinition tableDefinition);

		bool CreateIndex(TableDefinition tableDefinition, Index index);
		bool DropIndex(TableDefinition tableDefinition, Index index);

		bool CreateColumn(TableDefinition tableDefinition, ColumnDefinition columnDefinition);
		bool DropColumn(TableDefinition tableDefinition, ColumnDefinition columnDefinition);

		bool BulkInsert(Records records, TableDefinition tableDefinition);
		bool CreateOrUpdate(Record record, TableDefinition tableDefinition);
		bool Create(Record record, TableDefinition tableDefinition);
		bool Update(Record record, TableDefinition tableDefinition);

		bool UpdateWithCriteria(Record record, TableDefinition tableDefinition, WhereClause whereClause);
		bool UpdateCustom(string customerSql);
		bool Delete(string id, TableDefinition tableDefinition);
		bool DeleteAll(TableDefinition tableDefinition);

		long GetCount(TableDefinition tableDefinition);

		void GetManyStream(int n, TableDefinition tableDefinition, IProtoStreamer streamer);

		Record Get(string id, TableDefinition tableDefinition);
		Records GetCustom(string customSql, TableDefinition tableDefinition);
		Records GetMany(int n, TableDefinition tableDefinition);
		Records Where(WhereClause whereClause, TableDefinition tableDefinition);
	}
}

